File: 141_ConstructTreePreorderInorder.py
"""
Problem: Construct Binary Tree from Preorder and Inorder
Description: Build binary tree from preorder and inorder traversal arrays.

Approach: First element of preorder is root. Find root in inorder to split
          into left and right subtrees. Recursively build left and right.
"""

from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def buildTreeHelper(preorder: List[int], preStart: int, preEnd: int,
                    inorder: List[int], inStart: int, inEnd: int,
                    inMap: dict) -> Optional[TreeNode]:
    if preStart > preEnd or inStart > inEnd:
        return None
    
    rootVal = preorder[preStart]
    root = TreeNode(rootVal)
    
    inRoot = inMap[rootVal]
    numsLeft = inRoot - inStart
    
    root.left = buildTreeHelper(preorder, preStart + 1, preStart + numsLeft,
                                inorder, inStart, inRoot - 1, inMap)
    root.right = buildTreeHelper(preorder, preStart + numsLeft + 1, preEnd,
                                 inorder, inRoot + 1, inEnd, inMap)
    
    return root

def buildTree(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    inMap = {val: i for i, val in enumerate(inorder)}
    
    return buildTreeHelper(preorder, 0, len(preorder) - 1,
                          inorder, 0, len(inorder) - 1, inMap)

def printInorder(root: TreeNode):
    if not root:
        return
    printInorder(root.left)
    print(root.val, end=" ")
    printInorder(root.right)

def main():
    preorder = [3, 9, 20, 15, 7]
    inorder = [9, 3, 15, 20, 7]
    
    root = buildTree(preorder, inorder)
    
    print("Inorder traversal:", end=" ")
    printInorder(root)
    print()

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""

File: 141_ConstructTreePreorderInorder.cpp
/*
Problem: Construct Binary Tree from Preorder and Inorder
Description: Build binary tree from preorder and inorder traversal arrays.

Approach: First element of preorder is root. Find root in inorder to split
          into left and right subtrees. Recursively build left and right.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTreeHelper(vector<int>& preorder, int preStart, int preEnd,
                          vector<int>& inorder, int inStart, int inEnd,
                          unordered_map<int, int>& inMap) {
    if (preStart > preEnd || inStart > inEnd)
        return nullptr;
    
    int rootVal = preorder[preStart];
    TreeNode* root = new TreeNode(rootVal);
    
    int inRoot = inMap[rootVal];
    int numsLeft = inRoot - inStart;
    
    root->left = buildTreeHelper(preorder, preStart + 1, preStart + numsLeft,
                                 inorder, inStart, inRoot - 1, inMap);
    root->right = buildTreeHelper(preorder, preStart + numsLeft + 1, preEnd,
                                  inorder, inRoot + 1, inEnd, inMap);
    
    return root;
}

TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    unordered_map<int, int> inMap;
    for (int i = 0; i < inorder.size(); i++) {
        inMap[inorder[i]] = i;
    }
    
    return buildTreeHelper(preorder, 0, preorder.size() - 1,
                          inorder, 0, inorder.size() - 1, inMap);
}

void printInorder(TreeNode* root) {
    if (!root) return;
    printInorder(root->left);
    cout << root->val << " ";
    printInorder(root->right);
}

int main() {
    vector<int> preorder = {3, 9, 20, 15, 7};
    vector<int> inorder = {9, 3, 15, 20, 7};
    
    TreeNode* root = buildTree(preorder, inorder);
    
    cout << "Inorder traversal: ";
    printInorder(root);
    cout << endl;
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/
